cleverhans/feat_advs.py

```diff
@@ -0,0 +1,151 @@
+        self.feedable_kwargs = {'eps': np.float32,
+                                'eps_iter': np.float32,
+                                'clip_min': np.float32,
+                                'clip_max': np.float32,
+                                'layer': str}
+        self.structural_kwargs = ['ord', 'nb_iter']
+
+        assert isinstance(self.model, Model)
+
+    def parse_params(self, layer=None, eps=0.3, eps_iter=0.05, nb_iter=10,
+                     ord=np.inf, clip_min=None, clip_max=None, **kwargs):
+        """
+        Take in a dictionary of parameters and applies attack-specific checks
+        before saving them as attributes.
+
+        Attack-specific parameters:
+        :param layer: (required str) name of the layer to target.
+        :param eps: (required float) maximum distortion of adversarial example
+                    compared to original input
+        :param eps_iter: (required float) step size for each attack iteration
+        :param nb_iter: (required int) Number of attack iterations.
+        :param ord: (optional) Order of the norm (mimics Numpy).
+                    Possible values: np.inf, 1 or 2.
+        :param clip_min: (optional float) Minimum input component value
+        :param clip_max: (optional float) Maximum input component value
+        """
+
+        # Save attack-specific parameters
+        self.layer = layer
+        self.eps = eps
+        self.eps_iter = eps_iter
+        self.nb_iter = nb_iter
+        self.ord = ord
+        self.clip_min = clip_min
+        self.clip_max = clip_max
+
+        # Check if order of the norm is acceptable given current implementation
+        if self.ord not in [np.inf, 1, 2]:
+            raise ValueError("Norm order must be either np.inf, 1, or 2.")
+        if self.back == 'th':
+            error_string = ("FeatureAdversaries is"
+                            " not implemented in Theano")
+            raise NotImplementedError(error_string)
+
+        return True
+
+    def attack_single_step(self, x, eta, g_feat):
+        """
+        TensorFlow implementation of the Fast Feature Gradient. This is a
+        single step attack similar to Fast Gradient Method that attacks an
+        internal representation.
+
+        :param x: the input placeholder
+        :param eta: A tensor the same shape as x that holds the perturbation.
+        :param g_feat: model's internal tensor for guide
+        :return: a tensor for the adversarial example
+        """
+        from cleverhans.utils_tf import clip_eta
+
+        adv_x = x + eta
+        a_feat = self.model.get_layer(adv_x, self.layer)
+
+        # feat.shape = (batch, c) or (batch, w, h, c)
+        axis = range(1, len(a_feat.shape))
+
+        # Compute loss
+        # This is a targeted attack, hence the negative sign
+        loss = -tf.reduce_sum(tf.square(a_feat - g_feat), axis)
+
+        # Define gradient of loss wrt input
+        grad, = tf.gradients(loss, x)
```

Comment:
Shouldn't this be `adv_x` here instead of `x`?

Reply:
Thanks for catching that. Fixed in 11919a2.